Add ReverseRegistrarV2

stevieraykatz opened the pull request:
To fully support ENSIP-19,  two data migrations are needed: 
1. Our users' fwd resolution data needs to be set against the base network-specific cointype (per ENSIP-11). 
2. Our users' reverse resolution data needs to be stored in the ENS deployed L2ReverseRegistrar.

To address number 1, an owner-only hook has been added which allows the basenames admin to write ENSIP-11 data on behalf of existing basenames. 
 
To address number 2 and to ensure data is not lost in the migration process, the reverse data will be written during the migration period to both the legacy resolver and to the new ENS contract.

cb-heimdall commented:
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>
<!-- CODEOWNERS & Bar Raisers -->
<!-- Power Rangers -->
<!-- Links -->

stevieraykatz commented:
#123 now depends on this contract since we have wrapped the 2-step reverse set in a single call in reverse registrar

ilikesymmetry commented on src/L2/ReverseRegistrarV2.sol:
```diff
@@ -0,0 +1,320 @@
+    ///     If any of these checks fails, we skip this node and continue.
+    ///
+    /// @param nodes The array of nodes for which records will be set.
+    function setBaseForwardAddr(bytes32[] memory nodes) public onlyOwner {
```
security checks make sense to me but thinking about the offchain devx of `continue` vs event emission. 

For the last check, seems like we still reach the desired end-state which is good. 

For second to last, it seems like the name no longer goes anywhere so a reverse resolution wouldn't even validate according to ENSIP-19 so this is fine.

For first check, I'm not sure what resolver a user would switch to, but our hands are tied if this happens so I guess we couldn't even retry if we wanted to.

Mainly thinking from the perspective of having a big script churning through transactions and wanting to make it easy to not accidentally retry a node that didn't work for some reason, but I think there are purely offchain methods we can use so this looks good.

stevieraykatz replied:
I did consider an event-based approach (emitting before continuing) but thought that with offchain validation we can ensure that each call we make to this method will do what we want for each valid case. 

For indexing purposes the `setAddr` call will emit the only required event. 

New events would be for purely logging/script-runner convenience and I think we have enough other information to confirm the state of the changes without adding execution gas and state bloat.

ilikesymmetry replied:
aligned!

ilikesymmetry commented on test/ReverseRegistrarV2/SetNameForAddrWithSignature.t.sol:
```diff
@@ -0,0 +1,35 @@
+import {Sha3} from "src/lib/Sha3.sol";
+
+contract SetNameForAddrWithSignature is ReverseRegistrarV2Base {
+    function test_setsNameForAddrWithSignature() public {
```
Would prefer fuzzing on tests. In this case, seems like we could fuzz most args and generate signatures with a util to validate. Ideally we could also have an integration test against the deployed L2ReverseRegistrar but that doesn't exist yet right? For now maybe just having some basic fuzzing would be nice.

stevieraykatz replied:
Though I generally agree that fuzzing should be applied liberally, I think fuzzing around signature validation is out of scope for this contact. We rely on an external contract to do the signature validation so it doesn't make sense to me to add that validation to this repo's scope.

I agree an integration test would be ideal. We can technically do this against testnet today (with an older version) but I think a more valid integration test would use mainnet's deployment when it goes live. Thoughts on the specifics here?

ilikesymmetry reviewed (commented):
Couldn't find test for `setBaseForwardAddr`?

stevieraykatz commented:
> Couldn't find test for `setBaseForwardAddr`?

Added tests.

Also found a bug where the helper `_bytesToAddress` reverts for the zero address. Changed the third validation logic step in `setBaseForwardAddr`.

ilikesymmetry commented on src/L2/ReverseRegistrarV2.sol:
```diff
@@ -0,0 +1,321 @@
+    ///     https://github.com/ensdomains/ens-contracts/blob/staging/contracts/resolvers/profiles/AddrResolver.sol
+    ///
+    /// @param b Address bytes.
+    function _bytesToAddress(bytes memory b) internal pure returns (address payable a) {
+        require(b.length == 20);
+        assembly {
+            a := div(mload(add(b, 32)), exp(256, 12))
+        }
+    }
```
@stevieraykatz aligned with this github flag if we don't plan to use this util for anything

stevieraykatz replied:
Good catch slither! I just removed the last instance of it in the bug fix

ilikesymmetry reviewed (approved):
lgtm, would appreciate noting somewhere (maybe a linear ticket) that we should follow up with integration tests once ENS has stable bytecode.

stevieraykatz commented:
> lgtm, would appreciate noting somewhere (maybe a linear ticket) that we should follow up with integration tests once ENS has stable bytecode.

https://linear.app/coinbase/issue/BA-1952 added to track
